File: lib/services/integrations/integration_command_parser.dart
/// Resultado del parseo de un comando `/folio …` (espejo del servidor).
sealed class IntegrationCommandParseResult {
  const IntegrationCommandParseResult();
}

class IntegrationCommandLink extends IntegrationCommandParseResult {
  const IntegrationCommandLink(this.code);
  final String code;
}

class IntegrationCommandCreateTask extends IntegrationCommandParseResult {
  const IntegrationCommandCreateTask(this.title);
  final String title;
}

class IntegrationCommandListTasks extends IntegrationCommandParseResult {
  const IntegrationCommandListTasks();
}

class IntegrationCommandCompleteTask extends IntegrationCommandParseResult {
  const IntegrationCommandCompleteTask(this.title);
  final String title;
}

class IntegrationCommandUnknown extends IntegrationCommandParseResult {
  const IntegrationCommandUnknown();
}

/// Normaliza y parsea texto de comando Slack/Teams/Discord.
class IntegrationCommandParser {
  const IntegrationCommandParser();

  static const int maxTaskTitleLength = 200;

  IntegrationCommandParseResult parse(String raw) {
    var text = raw.trim();
    text = text.replaceFirst(RegExp(r'^<at>.*?</at>\s*', caseSensitive: false), '');
    text = text.replaceFirst(RegExp(r'^@\S+\s+'), '');
    text = text.trim();

    final link = RegExp(r'^/folio\s+link\s+([A-Za-z0-9]{6,12})$', caseSensitive: false)
        .firstMatch(text);
    if (link != null) {
      return IntegrationCommandLink(link.group(1)!.toUpperCase());
    }

    final create = RegExp(
      r'^/folio\s+create\s+task\s+"([^"]+)"$',
      caseSensitive: false,
    ).firstMatch(text);
    if (create != null) {
      final title = create.group(1)!.trim();
      if (title.isEmpty || title.length > maxTaskTitleLength) {
        return const IntegrationCommandUnknown();
      }
      return IntegrationCommandCreateTask(title);
    }

    final listTasks = RegExp(
      r'^/folio\s+list\s+tasks\s*$',
      caseSensitive: false,
    ).firstMatch(text);
    if (listTasks != null) {
      return const IntegrationCommandListTasks();
    }

    final complete = RegExp(
      r'^/folio\s+(?:complete\s+task|done)\s+"([^"]+)"$',
      caseSensitive: false,
    ).firstMatch(text);
    if (complete != null) {
      final title = complete.group(1)!.trim();
      if (title.isEmpty || title.length > maxTaskTitleLength) {
        return const IntegrationCommandUnknown();
      }
      return IntegrationCommandCompleteTask(title);
    }

    return const IntegrationCommandUnknown();
  }
}

File: backend/src/main/java/com/folio/backend/integrations/IntegrationCommandParser.java
package com.folio.backend.integrations;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class IntegrationCommandParser {

  private static final int MAX_TASK_TITLE_LEN = 200;
  private static final Pattern LINK =
      Pattern.compile("^/folio\\s+link\\s+([A-Za-z0-9]{6,12})$", Pattern.CASE_INSENSITIVE);
  private static final Pattern CREATE =
      Pattern.compile("^/folio\\s+create\\s+task\\s+\"([^\"]+)\"$", Pattern.CASE_INSENSITIVE);
  private static final Pattern LIST =
      Pattern.compile("^/folio\\s+list\\s+tasks\\s*$", Pattern.CASE_INSENSITIVE);
  private static final Pattern COMPLETE =
      Pattern.compile(
          "^/folio\\s+(?:complete\\s+task|done)\\s+\"([^\"]+)\"$", Pattern.CASE_INSENSITIVE);

  private IntegrationCommandParser() {}

  public static String normalizeCommandText(String raw) {
    String t = raw == null ? "" : raw.trim();
    t = t.replaceAll("(?i)^<at>.*?</at>\\s*", "");
    t = t.replaceAll("^@\\S+\\s+", "");
    return t.trim();
  }

  public sealed interface Parsed
      permits Parsed.Link, Parsed.CreateTask, Parsed.ListTasks, Parsed.CompleteTask, Parsed.Unknown {
    record Link(String code) implements Parsed {}

    record CreateTask(String title) implements Parsed {}

    record ListTasks() implements Parsed {}

    record CompleteTask(String title) implements Parsed {}

    record Unknown() implements Parsed {}
  }

  public static Parsed parse(String raw) {
    String text = normalizeCommandText(raw);
    Matcher link = LINK.matcher(text);
    if (link.matches()) {
      return new Parsed.Link(link.group(1).toUpperCase());
    }
    Matcher create = CREATE.matcher(text);
    if (create.matches()) {
      String title = create.group(1).trim();
      if (title.isEmpty() || title.length() > MAX_TASK_TITLE_LEN) {
        return new Parsed.Unknown();
      }
      return new Parsed.CreateTask(title);
    }
    if (LIST.matcher(text).matches()) {
      return new Parsed.ListTasks();
    }
    Matcher complete = COMPLETE.matcher(text);
    if (complete.matches()) {
      String title = complete.group(1).trim();
      if (title.isEmpty() || title.length() > MAX_TASK_TITLE_LEN) {
        return new Parsed.Unknown();
      }
      return new Parsed.CompleteTask(title);
    }
    return new Parsed.Unknown();
  }
}
